dsa-23au/DSAFormat.md

```diff
@@ -0,0 +1,231 @@
+algorithm (any functions currently defined and steppable through)
+and any data structures it is operating on.
+
+```
+Step
+count: <step_count>
+nextFunc: <id of function at end of this step>
+nextLine: <line number of function at end of this step>
+returnVal: <value to assign variable on current line>
+l: 3 
+<id1 of ds>: highlight <index1> # highlight the value at index1 
+<id2 of ds>: value <index2> <value2> # change the value at index2 to value2
+<id3 of ds>: highlight <index3> value <index4> <value4>
+```
+### Assignment Statement 
+
+Adds a new variable called `var1` to the function scope, initialized to `value1`.
+
+```
+var1 <-- value 
+```
+
+### Function Call 
+
+```
+var1, var2 <-- call <function id> arg1, arg2, ...
+```
+
+### If Statement 
+
+```
+if <expression1> 
+    <body>
+end
+```
+
+### While Loop 
+
+```
+while <expression>
+```
+
+### Do While Loop
+
+Begins a `while` loop which executes at least once, and evaluates its continue condition.
+
+```
+do 
+    <body>
+while <expression>
+```
+
+### End Code block
+
+Ends a code block begun by an `if` statement or a `while / for` loop.
+
+## Data Structures
+
+There are x built-in data structures in this first version, and more may be added
+in the future. Each can be oriented "Down", meaning drawn vertically with increasing indices
+going down, or "Right", meaning draw horizontally with increasing indices going to the right.
+
+```
+<ds type>
+id: zzz
+l: xx
+n: yyy
+<body>
+...
+... <end-of-data after <l> lines of body>
+```
+
+### Arrays
+
+Our most fundamental data structure, underlying all the others and allowing
+us to focus on just one memory range in our RAM model, is the array.
+
+We wish for it to label indices, and to contain references to objects of
+arbitrary types, not just integers as often appear in DSA textbooks.
+
+All object references have names beginning with `$` followed by a unique abbreviation
+of their identifier on the server (in Java, it is the memory address of the object on the heap).
+
+#### 1D Arrays
+
+```
+Array
+id: <id>
+n: 143
+dim: 1
+l1: 3
+$af $b3 $12 ...                                                              | 
+$g5 $b3 $12 ...                                                              | 
+$af $83 $7z
+refs: 99
+$af: <dev.codewithfriends.ClassA {a: 1, b:2}>
+$b3: <java.lang.Integer {3}>
+$12: <java.util.Map.Entry {k: $b3, v: $af}>
+```
+
+#### 2D Arrays
+
+```
+Array
+id: <id>
+n: 143
+dim: 2
+l1: 3
+$af $b3 $12 ...                                                              | 
+$g5 $b3 $12 ...                                                              | 
+$af $83 $7z
+l2: 2 
+$af $b3 $12 ...                                                              | 
+$g5 $b3 $12 ...                                                              | 
+$af $83 $7z
+refs: 182
+$af: <dev.codewithfriends.ClassA {a: 1, b:2}>
+$b3: <java.lang.Integer {3}>
+$12: <java.util.Map.Entry {k: $b3, v: $af}>
+```
+
+## Linked List
```

Comment:
could a linked list possibly be better?

Reply:
Let's keep thinking if linked lists might be useful for Spotify data. The use case might look like:
* order matters, like your listening preferences at ages 10, 15, 20, 25
* new updates keep arriving every day, that need to be easily inserted / removed without shifting data in an array